vllm/worker/worker.py

```diff
@@ -152,8 +152,47 @@ def execute_model(
         if cache_events is not None:
             for event in cache_events:
                 event.wait()
+
+    @torch.inference_mode()
+    def execute_model(
+        self,
+        seq_group_metadata_list: Optional[List[SequenceGroupMetadata]] = None,
+        blocks_to_swap_in: Optional[Dict[int, int]] = None,
+        blocks_to_swap_out: Optional[Dict[int, int]] = None,
+        blocks_to_copy: Optional[Dict[int, List[int]]] = None,
+    ) -> Optional[SamplerOutput]:
+        if self.is_driver_worker:
+            assert seq_group_metadata_list is not None
+            num_seq_groups = len(seq_group_metadata_list)
+            assert blocks_to_swap_in is not None
+            assert blocks_to_swap_out is not None
+            assert blocks_to_copy is not None
+            # Turn the dictionaries into lists for communication.
+            swap_in_src = list(blocks_to_swap_in.keys())
+            swap_in_dst = list(blocks_to_swap_in.values())
+            swap_out_src = list(blocks_to_swap_out.keys())
+            swap_out_dst = list(blocks_to_swap_out.values())
+            copy_src = []
+            copy_dst = []
+            for src, dst_list in blocks_to_copy.items():
+                copy_src.extend([src] * len(dst_list))
+                copy_dst.extend(dst_list)
+            swapping_block_numbers = [
+                swap_in_src, swap_in_dst, swap_out_src, swap_out_dst, copy_src,
+                copy_dst
+            ]
+            broadcast_object_list([num_seq_groups] + swapping_block_numbers,
+                                  src=0)
+        else:
+            recv_data = [None] * 7
+            broadcast_object_list(recv_data, src=0)
+            num_seq_groups = recv_data[0]
+            swapping_block_numbers = recv_data[1:]
```

Comment:
This pattern is repeated. Can we somehow factor this out?

Reply:
Do you mean factoring the following part out?

```python
if self.is_driver_worker:
    # some manual serialization logic to get the python object and 
    # the tensors to broadcast
    ...
    # broadcast the python object
    torch.distributed.broadcast_object_list([python_object], src=0)
    # broadcast the tensors
    torch.distributed.broadcast(..., src=0)
else:
    # receive the python object
    python_object = torch.distributed.broadcast_object_list(
        [None], src=0)[0]
    # receive the tensors
    torch.distributed.broadcast(..., src=0)
```

I can give it a try, but the potential issue is that the manual serialization logic can be pretty arbitrary.